### ISSUE TITLE
Is there a way for starting background tasks just after starting Orchard?

### ISSUE THREAD

**Issue by jersiovic** [OrchardCMS/OrchardCore#10582]
### Is your feature request related to a problem? Please describe.
My problem is that when an orchard app is started, I need to make a request to parent tenant and to every child tenant to be sure bakground tasks of weach tenant start to run. I don't want this is needed.

### Describe the solution you'd like
I would like Orchard itself starts its tenants and its background tasks without the need of a reques for each one of them

### Describe alternatives you've considered
An alternative is on start to make programmatically a request for each tenant.

**Comment by jtkech** [comment]
@jersiovic 

Yes, it is intended to lazily build a given tenant on a first request, which is required to run its bg tasks.

Would it be enough to only have to do a request on the default tenant? If so a dedicated middleware registered by the default tenant could manage it (maybe based on a feature that would need to be enabled). As a first direction we can look at the AutoSetup middleware.

Hmm, otherwise an hosted service that would manage this, maybe a config option to validate this beharvior OR just registering or not this hosted service would be enough.

**Comment by jersiovic** [comment]
Thank you I will explore hosted service path

**Comment by Skrypt** [comment]
You could probably also create your own hosted background task if there is no need to wake up the tenant. But in case that what you want is to execute a tenant with its background tasks then there is no choice but to wake up the tenant. This makes me think that we should have a setting per tenant to keep them alive at all costs. Sometimes, you might want to keep the tenant alive because it is crucial that these background tasks execute periodically.

**Comment by jersiovic** [comment]
In my case I don't need to wake up the tenant because if there's not any request it is ok if it is not wake up.
But I need its services per tenant, for accessing its data and its settings in teh background task

**Comment by jtkech** [comment]
> In my case I don't need to wake up the tenant because if there's not any request it is ok if it is not wake up.

Okay

> But I need its services per tenant, for accessing its data and its settings in the background task

Its services / data? services / data of the default tenant or any tenant?
in the background task? A bg task run by the default tenant or any tenant?

Not sure I understand, can you rephrase it?

For now what I can say is that for any tenant (including the default one) you need a first request on it (both the shell and its pipeline need to be built) to run its bg tasks. But when a bg task of any tenant (including the default one) runs, it can use the services of any other tenant by getting a scope on it (the related shell will be built if not yet done).

**Comment by Skrypt** [comment]
What I'm asking myself, to be more specific; is it necessary to host a tenant to access its services? Or is it that we need to host a tenant because some of these services are also middlewares?

**Comment by jtkech** [comment]
When we start an app our hosted bg service (host level so above any tenant) starts and wait for any tenant to be built including its pipeline, then at least we need a first request on any tenant to init the shells (without fully building them) and we build the shell (which calls the module startup ConfigureServices()) and the pipeline (which calls the module startup Configure())  of the tenant on which the request has been done.

Buiding the pipeline mainly means configure the routes and the middlewares of a given tenant, which is related to things to be done when we do a request on this tenant, e.g. for authorisation, routing, get static files and so on.

We can use the services (not related to a request) of a given tenant / shell even if its pipeline is not yet built, we only need to build the shell container (which call the module ConfigureServices()), but to save resources we opted to lazily build a given tenant only on a first request on it, not to fully build all tenants on the first request on any tenant.

That's why in the bg service we wait for a given tenant to be fully built including its pipeline meaning that a first request has been done on it. But we could have opted to fully build all shells on the first request on any tenant (currently they are only pre-created), in that case on the first request (we need at least one) on any tenant, all tenants would be built and then all bg tasks on any tenant would run (if we don't check anymore if the pipeline is built).

Hmm, maybe we could have a config options to build tenants lazily or not.

So it's doable to have options (maybe per tenant) to start bg tasks of all tenants with only one request on any tenant. And to do this first request automatically we would need an `IHostedService` that would do a request e.g. on the default tenant to warm up it and then build other tenants so that their bg tasks can run.

> What I'm asking myself, to be more specific; is it necessary to host a tenant to access its services?

We can access its services through REST api calls

> Or is it that we need to host a tenant because some of these services are also middleware?

A REST api call will do a request on the targetted tenant that will execute its middlewares

> What I'm thinking about is if it would be possible to create a separate host with a background task that would, just like we do when we start a tenant, retrieve the tenants' configurations then execute these background tasks within that scope.

Looks like another instance running a given tenant, we already can do this with a stateless config e.g. in a distributed environment.

> Now, what we do is instantiate background tasks on each of these hosted tenants because we need sometimes to pass into certain middleware to morph the data.

Middleware are important for results related to a given request, but normally bg tasks don't need them as they are not related to a given request, but a previous request may have change some states that a bg task can detect, and a bg task can initiate a request, there are so many scenarios ;) but what we can say is that a bg task don't execute in a request context (or a fake one), so here no middleware has to execute.

As said, in the bg service we wait for a given tenant to have its pipeline built, not because we need it but because we want to lazily build it on a first request on it (which will build its pipeline). Othewise we could get a scope on it before which will trigger a shell building if not yet done, so if a tenant has a bg task it would not be built lazily anymore, it would be built on the first execution of its bg task.

> So, you get the proper scoped results because you are within that hosted tenant context.

As said we can get scoped results without middlewares if we are talking about a shell scope, not a request scope context in which middlewares are executed.

> But what would we do if we needed to execute distributed background tasks?

Here many scenarios, e.g. the indexing bg task is executed on each instance but there are shared data in the shared database to know what to do, and each instance has a local file acting as a local cursor to be used on the shared indexing table. We can also use a distributed lock to execute a task atomically (one after each other) or exclusively across multiple running instances.

**Comment by Skrypt** [comment]
If we could configure tenants to be pre-warmed-up we could have one tenant that is always up and that would be dedicated for background tasks. Now, as you seem to say it would be possible that way to silently execute a background task on any tenant without warming it up.

**Comment by jersiovic** [comment]
> Its services / data? services / data of the default tenant or any tenant?
Data services and other services not dependant of the request but available at the tenant that have enabled the task. It could be the default tenant or a child tenant.
> in the background task? 
yes
>A bg task run by the default tenant or any tenant?
A bg task that can be run by default tenant or by a child tenant

**Comment by jersiovic** [comment]
> As said, in the bg service we wait for a given tenant to have its pipeline built, not because we need it but because we want to lazily build it on a first request on it (which will build its pipeline). Othewise we could get a scope on it before which will trigger a shell building if not yet done, so if a tenant has a bg task it would not be built lazily anymore, it would be built on the first execution of its bg task.

This sounds easiest solution and it will fit properly with our scenario

**Comment by jtkech** [comment]
@jersiovic 

Okay, so an app level configurable option (IConfiguration not IShellConfiguration) to check or not in the bg service if a given tenant pipeline is built, the default being to check it, if set to not check the pipeline we would only check if the tenant is in the running state.

Feel free to start a PR if you have time, otherwise I will do it when I will have time, ping me again if I forget ;)

**Comment by jersiovic** [comment]
Done. Thank you for your guidance
https://github.com/OrchardCMS/OrchardCore/pull/10638
